File: models/experimental/deit/tt/deit_attention.py
# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.

# SPDX-License-Identifier: Apache-2.0

import torch
from torch import nn
from typing import Union, Optional, Tuple


import ttnn

from models.experimental.deit.tt.deit_config import DeiTConfig
from models.experimental.deit.tt.deit_self_attention import TtDeiTSelfAttention
from models.experimental.deit.tt.deit_self_output import TtDeiTSelfOutput


class TtDeiTAttention(nn.Module):
    def __init__(self, config: DeiTConfig(), device, state_dict=None, base_address="") -> None:
        super().__init__()

        self.attention = TtDeiTSelfAttention(config, device, state_dict, f"{base_address}.attention")

        self.output = TtDeiTSelfOutput(config, device, state_dict, f"{base_address}.output")

    def forward(
        self,
        hidden_states: ttnn.Tensor,
        head_mask: Optional[ttnn.Tensor] = None,
        output_attentions: bool = False,
    ) -> Union[Tuple[ttnn.Tensor, ttnn.Tensor], Tuple[ttnn.Tensor]]:
        self_outputs = self.attention(hidden_states, head_mask, output_attentions)

        attention_output = self.output(self_outputs[0], hidden_states)

        outputs = (attention_output,) + self_outputs[1:]  # add attentions if we output them
        return outputs

File: models/experimental/deit/deit_cpp/tt_cpp/deit_attention.cpp
// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
//
// SPDX-License-Identifier: Apache-2.0

#include "deit_attention.h"
#include <stdexcept>

TtDeiTAttention::TtDeiTAttention(
    const DeiTConfig& config,
    std::shared_ptr<ttnn::MeshDevice> device,
    const std::unordered_map<std::string, torch::Tensor>& state_dict,
    const std::string& base_address
) : config(config), device(device) {
    if (!device) {
        throw std::invalid_argument("Device cannot be null");
    }

    // Create attention component with appropriate base address
    std::string attention_address = base_address.empty() ? "attention" : base_address + ".attention";
    attention = std::make_unique<TtDeiTSelfAttention>(config, device, state_dict, attention_address);

    // Create output component with appropriate base address
    std::string output_address = base_address.empty() ? "output" : base_address + ".output";
    output = std::make_unique<TtDeiTSelfOutput>(config, device, state_dict, output_address);
}

std::tuple<ttnn::Tensor, std::optional<ttnn::Tensor>> TtDeiTAttention::forward(
    const ttnn::Tensor& hidden_states,
    const std::optional<ttnn::Tensor>& head_mask,
    bool output_attentions
) {
    // Forward pass through self attention
    auto self_outputs = attention->forward(hidden_states, head_mask, output_attentions);
    
    // Extract attention output (first element of the tuple)
    ttnn::Tensor attention_hidden_states = std::get<0>(self_outputs);
    
    // Forward pass through output layer
    ttnn::Tensor attention_output = output->forward(attention_hidden_states, hidden_states);
    
    // Prepare return value
    if (output_attentions && std::get<1>(self_outputs).has_value()) {
        // Return both attention output and attention weights
        return std::make_tuple(attention_output, std::get<1>(self_outputs));
    } else {
        // Return only attention output
        return std::make_tuple(attention_output, std::nullopt);
    }
}